pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -20,33 +20,199 @@ import (
 	"k8s.io/api/core/v1"
 	k8stypes "k8s.io/apimachinery/pkg/types"
 	"k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vclib"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"strings"
+	"errors"
+	"sync"
 )
 
 // Stores info about the kubernetes node
 type NodeInfo struct {
+	dataCenter *vclib.Datacenter
 	vm       *vclib.VirtualMachine
 	vcServer string
 }
 
 type NodeManager struct {
+	// TODO: replace map with concurrent map when k8s supports go v1.9
+
 	// Maps the VC server to VSphereInstance
 	vsphereInstanceMap map[string]*VSphereInstance
 	// Maps node name to node info.
 	nodeInfoMap map[string]*NodeInfo
+	// Maps node name to node structure
+	registeredNodes map[string]*v1.Node
+
+
+	// Mutexes
+	registeredNodesLock sync.RWMutex
+	nodeInfoLock sync.RWMutex
+}
+
+const (
+	POOL_SIZE = 8
+	QUEUE_SIZE = POOL_SIZE * 10
+)
+
+var (
+	ErrNoNodeInfo = errors.New("no node info")
+	ErrNoVSphereInstance = errors.New("no vsphere instance found")
+)
+
+func (nm *NodeManager) DiscoverNode(node *v1.Node) {
+
+	type VmSearch struct {
+		vc string
+		datacenter *vclib.Datacenter
+	}
+
+	var queueChannel chan *VmSearch
+	var wg sync.WaitGroup
+	queueChannel = make(chan *VmSearch, QUEUE_SIZE)
+	nodeUUID := node.Status.NodeInfo.SystemUUID
+
+	go func() {
+		var datacenterObjs []*vclib.Datacenter
+		for vc, vsi := range nm.vsphereInstanceMap {
+			// Create context
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			err := vsi.conn.Connect(ctx)
+			if err != nil {
+				glog.V(4).Info("Discovering node error vc:", err)
+			}
+
+			if vsi.cfg.Datacenters == "" {
+				datacenterObjs, err = vclib.GetAllDatacenter(ctx, vsi.conn)
+				if err != nil {
+					glog.V(4).Info("Discovering node error dc:", err)
+				}
+			} else {
+				datacenters := strings.Split(vsi.cfg.Datacenters, ",")
+				for _, dc := range datacenters {
+					dc = strings.TrimSpace(dc)
+					if dc == "" {
+						continue
+					}
+					datacenterObj, err := vclib.GetDatacenter(ctx, vsi.conn, dc)
+					if err != nil {
+						glog.V(4).Info("Discovering node error dc:", err)
+					}
+					datacenterObjs = append(datacenterObjs, datacenterObj)
+				}
+			}
+
+			for _, datacenterObj := range datacenterObjs {
+				glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", node.Name, vc, datacenterObj.Name())
+				queueChannel <- &VmSearch{
+					vc:         vc,
+					datacenter: datacenterObj,
+				}
+			}
+		}
+		close(queueChannel)
+	}()
+
+	for i := 0; i < POOL_SIZE; i++ {
+		go func() {
```

Comment:
This go routine block waits for ever if the queueChannel is not populated at all by the time it starts executing this block. 
Instead you would want to wait until queueChannel is populated. This can be done without syncGroup. 
You can get <- from the channel queue. This will be a blocking call. After queueChannel is populated, you can search through all datacenters in parallel as sandeep suggested.

Reply:
The go routine will not block forever even if queueChannel is not populated. If the queue is empty, it gets unblocked when you close the channel. So, the logic wrt waiting infinitely is not present.

Reply:
Since we close channel this will not block forever

Reply:
Okay makes sense.